### DISCUSSION TITLE
[Help} Diarization does not detect host switching correctly

### DISCUSSION THREAD

**Discussion by Tribuzeus** [pyannote/pyannote-audio#1884] [Q&A]
Hi, I created a script with pyannote.audio, but it doesn't fast detect the speaker change very well.

[
  {
    "speaker_id": "SPEAKER_00",
    "start": 0.009,
    "end": 18.575
  },
  {
    "speaker_id": "SPEAKER_00",
    "start": 19.497,
    "end": 46.937
  },
  {
    "speaker_id": "SPEAKER_00",
    "start": 48.558,
    "end": 63.933
  },
  {
    "speaker_id": "SPEAKER_02",
    "start": 64.258,
    "end": 85.009
  },
  {
    "speaker_id": "SPEAKER_00",
    "start": 86.596,
    "end": 111.92
  },
  {
    "speaker_id": "SPEAKER_02",
    "start": 111.92,
    "end": 136.493
  },
  {
    "speaker_id": "SPEAKER_02",
    "start": 136.971,
    "end": 162.568
  },
  {
    "speaker_id": "SPEAKER_02",
    "start": 163.046,
    "end": 170.247
  }
] 
when the correct one would be something like this:
[{"speaker_id":"SPEAKER_00","start":0.008488964346349746,"end":1.3157894736842106},{"speaker_id":"SPEAKER_00","start":1.740237691001698,"end":4.320882852292021},{"speaker_id":"SPEAKER_00","start":4.609507640067912,"end":5.98471986417657},{"speaker_id":"SPEAKER_01","start":6.35823429541596,"end":7.071307300509337},{"speaker_id":"SPEAKER_01","start":7.34295415959253,"end":15.305602716468591},

my script:

import os
import json
import torch
import whisperx
from pyannote.audio import Pipeline
import pandas as pd

def listar_audios(pasta):
    return [f for f in os.listdir(pasta) if f.lower().endswith(('.wav', '.mp3', '.ogg', '.flac', '.m4a'))]

def annotation_to_df(annotation):
    data = []
    for segment, track, label in annotation.itertracks(yield_label=True):
        data.append({
            "start": segment.start,
            "end": segment.end,
            "speaker": str(label)
        })
    return pd.DataFrame(data)

def main():
    pasta = os.getcwd()
    arquivos = listar_audios(pasta)
    if not arquivos:
        print("Nenhum arquivo de áudio encontrado na pasta.")
        return

    print("Arquivos de áudio disponíveis:")
    for i, f in enumerate(arquivos, 1):
        print(f"{i}. {f}")
    idx = input("Digite o número do arquivo de áudio a ser usado: ").strip()
    try:
        idx = int(idx)
        audio_file = arquivos[idx-1]
    except:
        print("Entrada inválida.")
        return

    audio_path = os.path.join(pasta, audio_file)
    base_name = os.path.splitext(os.path.basename(audio_path))[0]
    folder = os.path.dirname(audio_path)
    json_path = os.path.join(folder, base_name + ".json")

    device_str = "cuda" if torch.cuda.is_available() else "cpu"
    device = torch.device(device_str)
    print(f"Carregando modelo WhisperX... (device: {device_str})")
    model = whisperx.load_model("medium", device_str, compute_type="float32")
    print("Transcrevendo áudio...")
    result = model.transcribe(audio_path, batch_size=1)
    
    print("Carregando pipeline de diarização pyannote...")
    diarization = Pipeline.from_pretrained("pyannote/speaker-diarization-3.1", use_auth_token=None).to(device)
    diarization_result = diarization(audio_path)
    
    print("Alinhando speakers...")
    diarization_df = annotation_to_df(diarization_result)
    diarized = whisperx.assign_word_speakers(diarization_df, result)

    json_output = []
    for segment in diarized["segments"]:
        json_output.append({
            "speaker_id": segment.get("speaker", "SPEAKER_UNKNOWN"),
            "start": segment["start"],
            "end": segment["end"]
        })

    print("Dados JSON:", json_output)
    print(f"Salvando JSON em: {json_path}")
    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(json_output, f, indent=2, ensure_ascii=False)
    print("Processo concluído!")

if __name__ == "__main__":
    main()

**Comment by arcticloud** [comment]
Two separate things are happening here.

**1. The granularity you want isn't diarization's job.** Your "correct" target has sub-second, phrase-level segments — that's the granularity of ASR word/phrase timestamps, not speaker diarization. Diarization returns *speaker turns* (who speaks when, in fairly coarse blocks) and no pre-trained diarizer will reproduce word-level boundaries. The usual pipeline is: run ASR (e.g. Whisper) to get word timings, run diarization to get speaker turns, then assign each word the speaker whose turn it falls in. That gives you fine segments *and* correct speaker labels.

**2. The missed host switches** are a merging/clustering issue. Levers, easiest first:

- If you know how many people speak, tell the pipeline — this is the single biggest fix:
  ```python
  diar = pipeline("audio.wav", num_speakers=2)
  # or, if unsure:  pipeline("audio.wav", min_speakers=2, max_speakers=4)
  ```
- Make sure your own post-processing isn't merging consecutive segments across a real switch (a common bug when you smooth/join same-label segments).
- For finer control you can re-instantiate the pipeline with a lower clustering `threshold` (default ~0.7; lower = more willing to split into distinct speakers) via `pipeline.instantiate({...})`.

Start with `num_speakers`/`min_speakers`/`max_speakers` — for a 2-host recording that alone usually cleans up the switching.
